src/lib/block_producer/block_producer.ml

```diff
@@ -140,14 +140,36 @@ let generate_next_state ~constraint_constants ~previous_protocol_state
         Consensus.Data.Block_data.coinbase_receiver block_data
       in
       let diff =
-        measure "create_diff" (fun () ->
-            Staged_ledger.create_diff ~constraint_constants staged_ledger
-              ~coinbase_receiver ~logger
-              ~current_state_view:previous_state_view
-              ~transactions_by_fee:transactions ~get_completed_work
-              ~log_block_creation ~supercharge_coinbase )
-        |> Result.map_error ~f:(fun err ->
-               Staged_ledger.Staged_ledger_error.Pre_diff err )
+        let diff =
+          measure "create_diff" (fun () ->
+              Staged_ledger.create_diff ~constraint_constants staged_ledger
+                ~coinbase_receiver ~logger
+                ~current_state_view:previous_state_view
+                ~transactions_by_fee:transactions ~get_completed_work
+                ~log_block_creation ~supercharge_coinbase )
+          |> Result.map_error ~f:(fun err ->
+                 Staged_ledger.Staged_ledger_error.Pre_diff err )
+        in
+        match diff with
+        | Ok d when Option.is_some block_reward_threshold ->
```

Comment:
nit: Pattern match on `block_reward_threshold` too and then you don't need the Option.value_exn later